File: atg1-optimized.java
import java.util.*;

public class Main {
    
    public static List<Integer> sortDescending(List<Integer> array) {
        array.sort(Collections.reverseOrder());
        return array;
    }

    public static boolean isGoodGraph(List<Integer> array) {
        return array.get(0) == 1 && Collections.frequency(array, 0) == array.size() - 1;
    }

    public static boolean hasNegativeElements(List<Integer> array) {
        for (int x : array) {
            if (x < 0) {
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        System.out.print("Zadaj pocet vrcholov: ");
        int numVertices = input.nextInt();
        List<Integer> degrees = new ArrayList<>(numVertices);
        for (int i = 0; i < numVertices; i++) {
            int degreeNum = i + 1;
            System.out.printf("Zadaj %d. stupen vrchola: ", degreeNum);
            degrees.add(input.nextInt());
        }
        List<Integer> sortedDegrees = sortDescending(degrees);
        System.out.println(sortedDegrees);
        for (int j = 0; j < sortedDegrees.size(); j++) {
            int temp = sortedDegrees.get(j);
            for (int k = 1; k <= temp; k++) {
                sortedDegrees.set(j, 0);
                sortedDegrees.set(j+k, sortedDegrees.get(j+k) - 1);
            }
            sortedDegrees = sortDescending(sortedDegrees);
            if (isGoodGraph(sortedDegrees)) {
                System.out.println("Uspesny graf");
                break;
            }
            if (hasNegativeElements(sortedDegrees)) {
                System.out.println("Neuspesny graf");
                break;
            }
        }
        input.close();
    }
}

File: atg1-optimized.py
def zorad(array):
    return sorted(array, reverse=True)

def existujeGraf(array):
    if array[0] != 1:
        return False
    for i in range(1, len(array)):
        if array[i] != 0:
            return False
    return True

def maZaporneHodnoty(array):
    return any(x < 0 for x in array)

def main():
    num_vertices = int(input("Zadaj pocet vrcholov: "))
    degrees = []
    for i in range(num_vertices):
        degree_num = i + 1
        degrees.append(int(input(f"Zadaj {degree_num}. stupen  vrchola: ")))
    sorted_degrees = zorad(degrees)
    print(sorted_degrees)
    for j in range(len(sorted_degrees)):
        temp = sorted_degrees[j]
        for k in range(1, temp+1):
            sorted_degrees[j] = 0
            sorted_degrees[j+k] -= 1
        sorted_degrees = zorad(sorted_degrees)
        if existujeGraf(sorted_degrees):
            print("Uspesny graf")
            break
        if maZaporneHodnoty(sorted_degrees):
            print("Neuspesny Graf")
            break

main()
